charger/pulsatrix.go

```diff
@@ -0,0 +1,269 @@
+package charger
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+	"nhooyr.io/websocket"
+)
+
+// pulsatrix charger implementation
+type PulsatrixCharger struct {
+	conn        *websocket.Conn
+	ctx         context.Context
+	uri         string
+	log         *util.Logger
+	enState     bool
+	bo          *backoff.ExponentialBackOff
+	signaledAmp float64
+	mutex       sync.Mutex
+	updated     time.Time
+	WebsocketData
+}
+
+type WebsocketData struct {
+	VehicleStatus     string    `json:"vehicleStatus"`
+	LastActivePower   float64   `json:"lastActivePower"`
+	PhaseVoltage      []float64 `json:"voltage"`
+	PhaseAmperage     []float64 `json:"amperage"`
+	AllocatedAmperage float64   `json:"allocatedAmperage"`
+	EnergyImported    float64   `json:"energyImported"`
+}
+
+func init() {
+	registry.Add("pulsatrix", NewPulsatrixFromConfig)
+}
+
+// NewPulsatrixtFromConfig creates a pulsatrix charger from generic config
+func NewPulsatrixFromConfig(other map[string]interface{}) (api.Charger, error) {
+	var cc struct{ Host string }
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+	return NewPulsatrix(cc.Host)
+}
+
+// NewPulsatrix creates pulsatrix charger
+func NewPulsatrix(hostname string) (*PulsatrixCharger, error) {
+	bo := backoff.NewExponentialBackOff()
+	bo.InitialInterval = 30 * time.Second
+	bo.MaxInterval = 5 * time.Minute
+	request.Timeout = 15 * time.Second
+	wb := PulsatrixCharger{
+		log:     util.NewLogger("pulsatrix"),
+		uri:     fmt.Sprintf("ws://%s/api/ws", hostname),
+		bo:      bo,
+		updated: time.Now(),
+	}
+
+	return &wb, wb.connectWs()
+}
+
+// ConnectWs connects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) connectWs() error {
+	ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
+	defer cancel()
+	c.ctx = ctx
+	c.log.INFO.Printf("connecting to %s", c.uri)
+	conn, _, err := websocket.Dial(ctx, c.uri, nil)
+	if err != nil {
+		if strErr := fmt.Sprintf("%s", err); strings.Contains(strErr, "context deadline exceeded") {
+			err = fmt.Errorf("Make sure the IP is correct and the SECC is connected to the network")
+		} else {
+			err = fmt.Errorf("error connecting to websocket: %v", err)
+		}
+		return err
+	}
+	c.updated = time.Now()
+	c.enState = false
+	c.conn = conn
+	c.handleError(c.Enable(false))
+	go c.wsReader()
+	go c.heartbeat()
+	return nil
+}
+
+// ReconnectWs reconnects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) reconnectWs() {
+	c.handleError(backoff.RetryNotify(func() error {
```

Comment:
```suggestion
	c.handleError(backoff.RetryNotify(c.connectWs),
```

Reply:
i am not sure what you mean with that. Is this what you wanted?
```suggestion
		notify := func(err error, time time.Duration) {
		c.log.WARN.Printf("trying to reconnect in %v...\n", time)
	}
	c.handleError(backoff.RetryNotify(c.connectWs, c.bo, notify)) 
	return
```

Reply:
You don't need the wrapper around `c.connectWs`, just remove it.